File: go/cmd/day01part2/main.go
package main

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/mlhoyt/aoc2020/go/pkg/utils"
)

func main() {
	input, err := utils.LoadInputFile("day01.txt")
	if err != nil {
		panic(err)
	}

	numbers, err := stringListToNumberList(strings.Split(input, "\n"))
	if err != nil {
		panic(err)
	}

	groups := numberListToNumberGroups(numbers)

	answers := filterNumberList(groups, func(p numberList) bool {
		return p[0]+p[1]+p[2] == 2020
	})

	for _, answer := range answers {
		fmt.Printf("%d * %d * %d = %d\n", answer[0], answer[1], answer[2], (answer[0] * answer[1] * answer[2]))
	}
}

type number int64

func stringListToNumberList(ss []string) ([]number, error) {
	ns := []number{}
	for _, s := range ss {
		n, err := strconv.ParseInt(s, 10, 64)
		if err != nil {
			return nil, err
		}

		ns = append(ns, number(n))
	}

	return ns, nil
}

type numberList []number

func newNumberList(ns ...number) numberList {
	return ns
}

func numberListToNumberGroups(ns []number) []numberList {
	results := []numberList{}
	for ai, a := range ns {
		for bi, b := range ns[(ai + 1):] {
			for _, c := range ns[(bi + 1):] {
				results = append(results, newNumberList(a, b, c))
			}
		}
	}

	return results
}

func filterNumberList(xs []numberList, f func(numberList) bool) []numberList {
	results := []numberList{}
	for _, x := range xs {
		if f(x) {
			results = append(results, x)
		}
	}

	return results
}

File: ts/src/day01part2/main.ts
import { readFileSync as fs_readFileSync } from 'fs';

// const FILE_NAME = "input/day01.test.txt";
const FILE_NAME = "input/day01.txt";

function main(): void {
  const input = fs_readFileSync(FILE_NAME, 'utf-8');
  // console.log("input: ", input);

  let numbers = input
    .split("\n")
    .map((s) => {return parseInt(s);});
  // console.log("numbers: ", numbers);

  let groups = generate_groups(numbers);
  // console.log("groups: ", groups);

  let answers = groups
    .filter(([x, y, z]) => { return (x + y + z) == 2020 });
  // console.log("answers: ", answers);

  answers.forEach(([x, y, z]) => { console.log(`${x} * ${y} * ${z}= ${x * y * z}`); });
}

function generate_groups(xs: number[]): [number, number, number][] {
    let groups = [];

    xs.forEach((x, xi) => {
        let ys = xs.slice(xi+1,);
        ys.forEach((y, yi) => {
            let zs = ys.slice(yi+1,);
            zs.forEach((z) => {
                groups.push([x, y, z]);
            });
        });
    })

    return groups
}

main()
